packages/light-client-extension-helpers/src/background/json-rpc-provider.ts

```diff
@@ -0,0 +1,205 @@
+     */
+    const messagePubSub = yield* PubSub.unbounded<string>()
+
+    /**
+     * An error pubsub for the daemon to publish errors to.
+     * It is the responsibility of the JSON RPC provider to consume these
+     *  errors and propagate them with the onError callback.
+     */
+    const errorPubSub = yield* PubSub.sliding<void>(1)
+
+    /**
+     * A synchronized reference to the current chain. This reference is updated
+     * every time the daemon restarts.
+     *
+     */
+    const chainRef = yield* $(
+      Effect.tryPromise(() => client.addChain(options)),
+      Effect.andThen(SynchronizedRef.make),
+    )
+
+    /**
+     * This daemon process listens for JSON RPC messages, publishing them to
+     * the message pubsub. It also detects errors and publishes those to the
+     * error pubsub.
+     *
+     * If an error occurs with `chain.nextJsonRpcResponse()`, the daemon
+     * restarts. This indicates that the chain is no longer functional.
+     * In response, the daemon reconnects to smoldot using the same
+     * `AddChainParameters` to obtain a new `Chain` object.
+     */
+    const daemon = yield* Effect.gen(function* () {
+      yield* Effect.addFinalizer(() => Queue.takeAll(daemonReadySignal))
+      yield* Effect.addFinalizer(() => PubSub.publish(errorPubSub, undefined))
+
+      const chain = yield* SynchronizedRef.updateAndGetEffect(
+        chainRef,
+        (oldChain) =>
+          $(
+            Effect.tryPromise(() => client.addChain(options)),
+            Effect.tap(() =>
+              $(
+                Effect.try(() => oldChain.remove()),
+                Effect.catchAll(() => Effect.void),
+              ),
+            ),
+          ),
+      )
+
+      // chain is initialized so we can signal that the daemon is ready
+      yield* Queue.offer(daemonReadySignal, undefined)
+
+      yield* $(
+        Effect.tryPromise(() => chain.nextJsonRpcResponse()),
+        Effect.tap(Effect.logDebug),
+        Effect.tap((message) => Queue.offer(messagePubSub, message)),
+      ).pipe(Effect.forever)
+    }).pipe(
+      // by scoping here, we ensure the finalizers are run before the next retry
+      Effect.scoped,
+      Effect.retry(Schedule.spaced("1 second")),
+      Effect.forkDaemon,
+      Effect.interruptible,
+    )
+
+    const provider: JsonRpcProvider = yield* Effect.sync(() =>
+      getSyncProvider(async () => {
+        await $(
+          Queue.take(daemonReadySignal),
+          Effect.tap(() => Effect.log("JSON RPC provider initialized.")),
+          runPromise,
+        )
+
+        return (onMessage, onError) => {
+          const invokeOnMessageCallback = (message: string) =>
+            $(
+              Effect.try(() => onMessage(message)),
+              Effect.catchAll(() => Effect.void),
+            )
+
+          const invokeOnErrorCallback = $(
+            Effect.try(() => onError()),
+            Effect.catchAll(() => Effect.void),
+          )
+
+          runCallback(
```

Comment:
why do we get hold of a runtime and call `runCallback` like this, instead of something like `Effect.run(..)` here? also below :)

Reply:
Its to ensure everything runs in the same context. If the application blew up somewhere (similar to a rust panic), whatever is running inside of `runCallback` & `runPromise` would immediately be cancelled. Whereas if you just ran it with `Effect.runPromise`, `Effect.runCallback`, it runs in its own context. 

Additionally, it perserves whatever you chose to bake into the runtime. So if I chose to set a custom logger, the effects that run inside of the `runPromise`, `runCallback` extracted from the runtime would use that logger instead of the default Effect or Console logger.